wakuv2/nwaku.go

```diff
@@ -2923,3 +2910,280 @@ func (self *Waku) DisconnectPeerById(peerId peer.ID) error {
+	requestIDStr := hex.EncodeToString(protocol.GenerateRequestID())
+
+	logger := s.logger.With(zap.Stringer("peerID", peerID), zap.String("requestID", requestIDStr))
+
+	logger.Debug("sending store request")
+
+	storeRequest := &storepb.StoreQueryRequest{
+		RequestId:         requestIDStr,
+		PubsubTopic:       proto.String(pubsubTopic),
+		ContentTopics:     contentTopics,
+		TimeStart:         from,
+		TimeEnd:           to,
+		IncludeData:       false,
+		PaginationCursor:  nil,
+		PaginationForward: false,
+		PaginationLimit:   proto.Uint64(pageSize),
+	}
+
+	jsonQuery, err := json.Marshal(storeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: timeouts need to be managed differently. For now we're using a 1m timeout
+	jsonResponse, err := wakuStoreQuery(s.wakuCtx, string(jsonQuery), peerID.String(), int(time.Minute.Milliseconds()))
+	if err != nil {
+		return nil, err
+	}
+
+	storeResponse := &storepb.StoreQueryResponse{}
+	err = json.Unmarshal([]byte(jsonResponse), storeResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	if storeResponse.GetStatusCode() != http.StatusOK {
+		return nil, fmt.Errorf("could not query storenode: %s %d %s", requestIDStr, storeResponse.GetStatusCode(), storeResponse.GetStatusDesc())
+	}
+
+	return newStoreResultImpl(s.wakuCtx, peerID, storeRequest, storeResponse), nil
+}
+
+type storeResultImpl struct {
+	done bool
+
+	wakuCtx       unsafe.Pointer
+	storeRequest  *storepb.StoreQueryRequest
+	storeResponse *storepb.StoreQueryResponse
+	peerID        peer.ID
+}
+
+func newStoreResultImpl(wakuCtx unsafe.Pointer, peerID peer.ID, storeRequest *storepb.StoreQueryRequest, storeResponse *storepb.StoreQueryResponse) *storeResultImpl {
+	return &storeResultImpl{
+		wakuCtx:       wakuCtx,
+		storeRequest:  storeRequest,
+		storeResponse: storeResponse,
+		peerID:        peerID,
+	}
+}
+
+func (r *storeResultImpl) Cursor() []byte {
+	return r.storeResponse.GetPaginationCursor()
+}
+
+func (r *storeResultImpl) IsComplete() bool {
+	return r.done
+}
+
+func (r *storeResultImpl) PeerID() peer.ID {
+	return r.peerID
+}
+
+func (r *storeResultImpl) Query() *storepb.StoreQueryRequest {
+	return r.storeRequest
+}
+
+func (r *storeResultImpl) Response() *storepb.StoreQueryResponse {
+	return r.storeResponse
+}
+
+func (r *storeResultImpl) Next(ctx context.Context, opts ...store.RequestOption) error {
+	// TODO: opts is being ignored. Will require some changes in go-waku. For now using this
+	// is not necessary
+
+	if r.storeResponse.GetPaginationCursor() == nil {
+		r.done = true
+		return nil
+	}
+
+	r.storeRequest.RequestId = hex.EncodeToString(protocol.GenerateRequestID())
+	r.storeRequest.PaginationCursor = r.storeResponse.PaginationCursor
+
+	jsonQuery, err := json.Marshal(r.storeRequest)
+	if err != nil {
+		return err
+	}
+
+	// TODO: timeouts need to be managed differently. For now we're using a 1m timeout
```

Comment:
What do you mean?

Reply:
In golang when we want to create a context with timeout, we do something like this:
```
ctx, cancel := context.WithTimeout(someContext, 10 * time.Second)

and then pass down the `ctx` to the function. The problem is that we do not have a way to read the timeout value that a context has, hence why i used a hardcoded value. Ideally we'd be able to obtain the timeout from the context, or make the function receive a timeout directly or maybe create our own context wrapper that would have a timeout accesor, but it's a decision we should take